src/core/lv_obj_class.c

```diff
@@ -165,22 +165,16 @@ bool lv_obj_is_group_def(lv_obj_t * obj)
  *   STATIC FUNCTIONS
  **********************/
 
-static void lv_obj_construct(lv_obj_t * obj)
+static void lv_obj_construct(lv_obj_t * obj, const lv_obj_class_t * class_p)
 {
-    const lv_obj_class_t * original_class_p = obj->class_p;
-
-    if(obj->class_p->base_class) {
+    if(class_p->base_class) {
         /*Don't let the descendant methods run during constructing the ancestor type*/
-        obj->class_p = obj->class_p->base_class;
 
         /*Construct the base first*/
-        lv_obj_construct(obj);
+        lv_obj_construct(obj, class_p->base_class);
     }
 
-    /*Restore the original class*/
-    obj->class_p = original_class_p;
-
-    if(obj->class_p->constructor_cb) obj->class_p->constructor_cb(obj->class_p, obj);
+    if(class_p->constructor_cb) class_p->constructor_cb(class_p, obj);
```

Comment:
I understand the obj->class_p needed to be changed during construction now. But we need to find a better way to call `lv_obj_assign_id`. Currently it's inside of basic `obj->constructor_cb`, where all widgets will go through. But withouth this change, obj->class_p is always `obj`.